org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/BaseConfigurator.java

```diff
@@ -0,0 +1,128 @@
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+
+import org.eclipse.buildship.core.GradleBuild;
+import org.eclipse.buildship.core.InitializationContext;
+import org.eclipse.buildship.core.ProjectConfigurator;
+import org.eclipse.buildship.core.ProjectContext;
+import org.eclipse.buildship.core.internal.CorePlugin;
+import org.eclipse.buildship.core.internal.marker.GradleErrorMarker;
+
+public class BaseConfigurator implements ProjectConfigurator {
+
+    private Map<File, EclipseProject> locationToProject;
+    private org.eclipse.buildship.core.internal.workspace.GradleBuild internalGradleBuild;
+
+    @Override
+    public void init(InitializationContext context, IProgressMonitor monitor) {
+        GradleBuild gradleBuild = context.getGradleBuild();
+        this.internalGradleBuild = ((org.eclipse.buildship.core.internal.DefaultGradleBuild)gradleBuild).getInternalGradleBuild();
+        Collection<EclipseProject> eclipseProjects = ModelProviderUtil.fetchAllEclipseProjects(this.internalGradleBuild, GradleConnector.newCancellationTokenSource(), FetchStrategy.LOAD_IF_NOT_CACHED, monitor);
+        this.locationToProject = eclipseProjects.stream().collect(Collectors.toMap(p -> p.getProjectDirectory(), p -> p));
+    }
+
+    @Override
+    public void configure(ProjectContext context, IProgressMonitor monitor) {
+        IProject project = context.getProject();
+        try {
+            configure(project, monitor);
+        } catch (CoreException e) {
+            GradleErrorMarker.create(context.getProject(), this.internalGradleBuild, "Failed to configure project", e, 0);
+            CorePlugin.logger().warn("Failed to configure project " + context.getProject().getName(), e);
+        }
+    }
+
+    private void configure(IProject project, IProgressMonitor monitor) throws CoreException {
+        SubMonitor progress = SubMonitor.convert(monitor);
+        progress.setWorkRemaining(4);
+
+        PersistentModelBuilder persistentModel = new PersistentModelBuilder(CorePlugin.modelPersistence().loadModel(project));
+        EclipseProject model = lookupEclipseModel(project);
+        progress.worked(1);
+
+        BuildScriptLocationUpdater.update(model, persistentModel, progress.newChild(1));
+
+        LinkedResourcesUpdater.update(project, ImmutableList.copyOf(model.getLinkedResources()), persistentModel, progress.newChild(1));
+        GradleFolderUpdater.update(project, model, persistentModel, progress.newChild(1));
+        ProjectNatureUpdater.update(project,  ImmutableList.copyOf(model.getProjectNatures()), persistentModel, progress.newChild(1));
+        BuildCommandUpdater.update(project, ImmutableList.copyOf(model.getBuildCommands()), persistentModel, progress.newChild(1));
+
+        if (isJavaProject(model)) {
```

Comment:
⭕️ Should probably be in its own Java configurator.

Reply:
I'll add a TODO for this and implement it as part of #786.